apps/src/lab2/views/ExtraLinks.tsx

```diff
@@ -14,61 +15,91 @@ interface PermissionResponse {
   permissions: string[];
 }
 
-interface ExtraLinksResponse {
-  links: {[key: string]: {text: string; url: string; access_key?: string}[]};
-  can_clone: boolean;
-  can_delete: boolean;
-  level_name: string;
-  script_level_path_links: {
-    script: string;
-    path: string;
-  }[];
+interface PermissionsAndLinkData {
+  levelLinkData: ExtraLinksLevelData | undefined;
+  projectLinkData: ExtraLinksProjectData | undefined;
+  permissions: string[];
+}
+
+async function fetchPermissionsAndLinkData(
+  levelId: number,
+  channelId?: string
+): Promise<PermissionsAndLinkData> {
+  // Fetch permissions.
+  const permissionsResponse = await HttpClient.fetchJson<PermissionResponse>(
+    '/api/v1/users/current/permissions'
+  );
+  const {permissions} = permissionsResponse.value;
+
+  // Fetch level link data.
+  let levelLinkData: ExtraLinksLevelData | undefined;
+  if (permissions.includes(PERMISSIONS.LEVELBUILDER)) {
+    const levelLinkDataResponse =
+      await HttpClient.fetchJson<ExtraLinksLevelData>(
+        `/levels/${levelId}/extra_links`
+      );
+    levelLinkData = levelLinkDataResponse.value;
+  }
+
+  // Fetch project link data.
+  let projectLinkData: ExtraLinksProjectData | undefined;
+  if (permissions.includes(PERMISSIONS.PROJECT_VALIDATOR)) {
+    const levelProjectDataResponse =
+      await HttpClient.fetchJson<ExtraLinksProjectData>(
+        `/projects/${channelId}/extra_links`
+      );
+    projectLinkData = levelProjectDataResponse.value;
+  }
+
+  // Return fetched permissions and link data.
+  return {
+    levelLinkData,
+    projectLinkData,
+    permissions,
+  };
 }
 
 // If the user has permission to see extra links, fetch extra links for the level,
 // then display a modal with the link data.
 const ExtraLinks: React.FunctionComponent<ExtraLinksProps> = ({
   levelId,
 }: ExtraLinksProps) => {
-  const {loading, data} = useFetch('/api/v1/users/current/permissions');
+  const [isLoading, setIsLoading] = useState(false);
+  const [PermissionsAndLinkData, setPermissionsAndLinkData] =
```

Comment:
nit: lowercase naming `permissionsAndLinkData` so it doesn't get confused with the type name?